HW1/P5/P5_CC.py

```diff
@@ -0,0 +1,159 @@
+from pyspark import SparkContext, SparkConf
+
+if 'sc' not in globals():
+	conf = SparkConf().setAppName('BFS')
+	sc = SparkContext(conf=conf)
+
+
+def swap_KV(KV):
+	return (KV[1],KV[0])
+
+def find_earliest(L):
+	if L[0]==None and L[1]==None:
+		return None
+	else:
+		return min(x for x in list(L) if x is not None)
+
+def update_rank_list(rank_list,new_char_list):
+	#	merges old rank list RDD=(CHAR,RANK) 
+	#	with new found characters RDD=(NEW_CHAR,NEW_RANK)
+	# 	from the current search
+	#	We do a join here, but the shuffle is pretty minimal, 
+	#	since everything should be
+	# 	copartitioned and the two RDDs are pretty small
+	
+	#	The left outer join will return a value of (x,y) where x is the old rank
+	#	and y is the new rank. If node has not been touched yet, x=None
+	#	If node is touched at this level, y=level
+	#	The mapValues function filters (x,y) to return either None, if 
+	#	node hasn't been touched even after this search level,
+	#or the lowest level found thus far
+	assert copartitioned(rank_list,new_char_list) #just to be sure
+	return rank_list.leftOuterJoin(new_char_list).mapValues(find_earliest)
```

Comment:
leftOuterJoin differs from a regular (inner join) by also keeping the extra entries in Left even if there is no match in the right side.  in terms of total time, it's probably inner join <= left join <= outer join